### PR TITLE
Refactor ManualPositionSmoothVel and ManualAltitudeSmoothVel

### PR DESCRIPTION
**Pull request by bresch**
With #12909 , we can get rid of a lot of corner case handling logic and I decided to refactor completely the FlightTaskManualPositionSmoothVel and FlightTaskManualAltitudeSmoothVel to make them cleaner, easier to read and bug-free.
To do that, I'll encapsulate the usage of the jerk-limited trajectory generator with the position lock logic into an object that can be used by both flight tasks. Being in a new class, I'll also have unit test for it.

First step I made so far is to split the logic in small functions and decouple XY and Z axes. Next step is to move that logic into a class. There will be a base pure virtual interface class, a class for XY and one for Z, where XY is the vector version of Z.

- [x] Extract XY and Z classes
- [x] Update PositionSmoothVel
- [x] Update AltitudeSmoothVel
- [x] Write unit tests (well.. I started at least)

### CODE REVIEW

**Comment by bresch** [comment]
Rebased on #12909 .

**Comment by bresch** [comment]
Rebased on master after the merge of #12909

File: src/systemcmds/tests/test_matrix.cpp [hunk 1]
```diff
@@ -578,7 +578,7 @@ bool MatrixTest::sliceTests()
 	};
 
 	Matrix<float, 2, 2> C(data_2);
-	A.set(C, 1, 1);
```
Comment (bresch) [inline src/systemcmds/tests/test_matrix.cpp]: FYI @jkflying 
See also the other modifications in this file.

**Comment by bresch** [comment]
@PX4/testflights Could you test this PR in -at least- attitude and position modes please?

- Altitude mode:
   - fly around, change altitude
   - up and down motion, wait sometimes a few seconds between motions
- Position control: 
   - fly around
   - change altitude
   - start and stop
   - hold position for 2-3 minutes

Thanks!

**Comment by bresch** [comment]
Status:
- [ ] waiting for #12964 to be merged first
- [x] tested by myself 
- [ ] tested by the PX4 test team

**Comment by dannyfpv** [comment]
tested on pixhawk4 v5 f-450

Altitude mode:
fly around, change altitude
up and down motion, wait sometimes a few seconds between motions

log:
https://review.px4.io/plot_app?log=80457d44-990b-4b4d-92d6-568fe7d11ea6


Position control:
fly around
change altitude
start and stop
hold position for 2-3 minutes

log:
https://review.px4.io/plot_app?log=993aa988-9843-4701-a6ac-2bb82478f970

**Comment by jorge789** [comment]
### Tested on CUAV+ V5:

Altitude mode:
fly around, change altitude
up and down motion, wait sometimes a few seconds between motions

**log:**
https://review.px4.io/plot_app?log=b6f62c4a-1260-468c-93fd-90e84d3b0dc3

Position control:
fly around
change altitude
start and stop
hold position for 2-3 minutes

**log:**
https://review.px4.io/plot_app?log=e8ffaa51-c4b6-4063-897c-662cf48f3382


### Tested on PixRacer V4:

Altitude mode:
fly around, change altitude
up and down motion, wait sometimes a few seconds between motions

**log:**
https://review.px4.io/plot_app?log=60fed4c6-4d91-436c-b636-3280853bc97c

Position control:
fly around
change altitude
start and stop
hold position for 2-3 minutes

**log:**
https://review.px4.io/plot_app?log=85e9741a-8572-464e-81ac-c303e72275dc

**Comment by bresch** [comment]
@dannyfpv @jorge789 Thanks, I'll have a look at the logs.

**Comment by bresch** [comment]
@MaEtUgR It's all ready now and tested quite a bit. The logs looks fine and I just need a quick review and approval please.

**Pull request merged**